{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Protocol for checking CDS offtargets:\n",
    "\n",
    "1. Get the augmented library guides (guides + 4 PAMs) and bowtie align them to CDS\n",
    "    1. This finds the appropriate NGG PAM guide variants\n",
    "    1. Two things can be observed from the alignment output:\n",
    "        1. For 0mm, which guide + PAM sequences have exact matches and their locations\n",
    "            1. If a guide has more than 1 exact match with the same or different PAM, those are off targets.\n",
    "        1. For 1mm, there are alignments that indicate 1 mismatch but in the N of the PAM region. I throw these out. Because these target the same region as the \"original\" guide + PAM with which they mismatch once. If they targeted another region with 1 mismatch, there would have been a correct guide + PAM variant in the previous step (0mm) which would have been recorded already.\n",
    "        1. For 2mm, there are alignments that indicate 1 mismatch in the N of the PAM, and 1 mismatch somewhere else in the sequence. I throw these out. Because the correct variant and 1 mismatch would have been recorded in the previous step (1mm)\n",
    "1. The number of exact matches of 0MM CDS must be the same as the library size.\n",
    "    1. Actually, size is a bad measure. What if the same guide with different PAMs has matches?\n",
    "1. Need to align augmented library to CDS, then add a column without PAMS. This column must have the same elements and size as the unaugmented library guides. The size of the alignment will be equal or more (w/ offtargets) than the library."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "if there is an intron separation event, we can't see it in the protocol above. Need to check genome.\n",
    "\n",
    "1. Align the augmented library to genome\n",
    "    1. This finds the appropriate NGG PAM guide variants.\n",
    "1. align augmented library to genome, then add a column without PAMS. This column must have the same elements and size as the unaugmented library guides. The size of the alignment will be:\n",
    "\n",
    "1. Equal\n",
    "    1. When either all guides have exactly 1 match (good!), or some guides have NO match (intron event) AND some have more than 1 match (offtarget event, double bad)\n",
    "1. Fewer\n",
    "    1. When some guides have no match (intron event)\n",
    "1. More\n",
    "    1. When some guides have more than 1 match (offtarget)\n",
    "\n",
    "than the library size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rc(string):\n",
    "    s = ''\n",
    "    for c in string:\n",
    "        if c == 'A': s += 'T'\n",
    "        elif c == 'C': s += 'G'\n",
    "        elif c == 'G': s += 'C'\n",
    "        elif c == 'T': s += 'A'\n",
    "    return s[::-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "def rc(string):\n",
    "    s = ''\n",
    "    for c in string:\n",
    "        if c == 'A': s += 'T'\n",
    "        elif c == 'C': s += 'G'\n",
    "        elif c == 'G': s += 'C'\n",
    "        elif c == 'T': s += 'A'\n",
    "    return s[::-1]\n",
    "\n",
    "\n",
    "def GO(species_name, track):\n",
    "    if not os.path.exists(f'{species_name}/output/{track}/cds_targets'):\n",
    "        os.makedirs(f'{species_name}/output/{track}/cds_targets')\n",
    "    \n",
    "    if not os.path.exists(f'{species_name}/output/{track}/genome_targets'):\n",
    "        os.makedirs(f'{species_name}/output/{track}/genome_targets')\n",
    "\n",
    "    library = pd.read_csv(f'{species_name}/{species_name}_{track}_library.csv')\n",
    "\n",
    "    zero_mm_cds_offtargets = [0] * len(library)\n",
    "    one_mm_cds_offtargets = [0] * len(library)\n",
    "    two_mm_cds_offtargets = [0] * len(library)\n",
    "\n",
    "    zero_mm_genome_offtargets = [0] * len(library)\n",
    "    one_mm_genome_offtargets = [0] * len(library)\n",
    "    two_mm_genome_offtargets = [0] * len(library)\n",
    "\n",
    "\n",
    "    df_a7_0mm_cds = pd.read_csv(f'{species_name}/output/{track}/{species_name}_{track}_0mm_cds_alignment.sam', sep='\\t',\n",
    "                names=['query_name', 'strand', 'reference_name', 'mapping_position',\n",
    "                       'aligned_seq', 'mapping_quality', 'idk', 'mismatch'])\n",
    "\n",
    "    df_a7_0mm_cds.drop(columns=['idk', 'mapping_quality'], inplace=True)\n",
    "\n",
    "    guide_w_pam = list()\n",
    "\n",
    "    # Add a reverse complement column if strand is negative - otherwise write the same sequence\n",
    "    for idx, row in df_a7_0mm_cds.iterrows():\n",
    "        if row['strand'] == '-':\n",
    "            guide_w_pam.append(rc(row['aligned_seq']))\n",
    "        else:\n",
    "            guide_w_pam.append(row['aligned_seq'])\n",
    "\n",
    "    df_a7_0mm_cds['guide_w_pam'] = guide_w_pam\n",
    "    df_a7_0mm_cds['guide'] = df_a7_0mm_cds['guide_w_pam'].str[:-3]\n",
    "\n",
    "    df_a7_0mm_cds['mismatch'].fillna('Exact Match', inplace=True)\n",
    "    df_a7_0mm_cds = df_a7_0mm_cds[~df_a7_0mm_cds['mismatch'].str.contains('20:')]\n",
    "    df_a7_0mm_cds['num_mutations'] = df_a7_0mm_cds['mismatch'].apply(lambda x: len(x.split(',')))\n",
    "    df_a7_0mm_cds.loc[df_a7_0mm_cds['mismatch'] == 'Exact Match', 'num_mutations'] = 0\n",
    "\n",
    "    df_a7_0mm_cds['num_mutations'] = df_a7_0mm_cds['num_mutations'].astype(int)\n",
    "\n",
    "    has_no_offtargets = (set(library['sequence']) == set(df_a7_0mm_cds['guide'])) and (len(library['sequence']) == len(df_a7_0mm_cds['guide']))\n",
    "\n",
    "    if not has_no_offtargets:\n",
    "        for idx, row in library.iterrows():\n",
    "            seq = row['sequence']\n",
    "            zero_mm_cds_offtargets[idx] += len(df_a7_0mm_cds[df_a7_0mm_cds['guide'] == seq]) - 1\n",
    "\n",
    "    ## -------\n",
    "    ## CDS 1MM\n",
    "    ## -------\n",
    "    df_a7_1mm_cds = pd.read_csv(f'{species_name}/output/{track}/{species_name}_{track}_1mm_cds_alignment.sam', sep='\\t',\n",
    "                names=['query_name', 'strand', 'reference_name', 'mapping_position',\n",
    "                       'aligned_seq', 'mapping_quality', 'idk', 'mismatch'])\n",
    "\n",
    "    df_a7_1mm_cds.drop(columns=['idk', 'mapping_quality'], inplace=True)\n",
    "\n",
    "    guide_w_pam = list()\n",
    "\n",
    "    for idx, row in df_a7_1mm_cds.iterrows():\n",
    "        if row['strand'] == '-':\n",
    "            guide_w_pam.append(rc(row['aligned_seq']))\n",
    "        else:\n",
    "            guide_w_pam.append(row['aligned_seq'])\n",
    "\n",
    "    df_a7_1mm_cds['guide_w_pam'] = guide_w_pam\n",
    "    df_a7_1mm_cds['guide'] = df_a7_1mm_cds['guide_w_pam'].str[:-3]\n",
    "\n",
    "    # Create a new column called num_mutations and extract the mismatches from\n",
    "    # the mismatch column.\n",
    "    df_a7_1mm_cds['mismatch'].fillna('Exact Match', inplace=True)\n",
    "    df_a7_1mm_cds = df_a7_1mm_cds[~df_a7_1mm_cds['mismatch'].str.contains('20:')]\n",
    "    df_a7_1mm_cds['num_mutations'] = df_a7_1mm_cds['mismatch'].apply(lambda x: len(x.split(',')))\n",
    "    df_a7_1mm_cds.loc[df_a7_1mm_cds['mismatch'] == 'Exact Match', 'num_mutations'] = 0\n",
    "\n",
    "    df_a7_1mm_cds['num_mutations'] = df_a7_1mm_cds['num_mutations'].astype(int)\n",
    "    \n",
    "    has_no_offtargets = (set(library['sequence']) == set(df_a7_1mm_cds['guide'])) and (len(library['sequence']) == len(df_a7_1mm_cds['guide']))\n",
    "\n",
    "    if not has_no_offtargets:\n",
    "        for _, row in df_a7_1mm_cds[df_a7_1mm_cds['num_mutations'] > 0].iterrows():\n",
    "            idx = library[library['sequence'] == row['guide']].index[0]\n",
    "            one_mm_cds_offtargets[idx] += 1\n",
    "\n",
    "    ## ------\n",
    "    # CDS 2MM\n",
    "    ## ------\n",
    "    df_a7_2mm_cds = pd.read_csv(f'{species_name}/output/{track}/{species_name}_{track}_2mm_cds_alignment.sam', sep='\\t',\n",
    "                names=['query_name', 'strand', 'reference_name', 'mapping_position',\n",
    "                       'aligned_seq', 'mapping_quality', 'idk', 'mismatch'])\n",
    "\n",
    "    df_a7_2mm_cds.drop(columns=['idk', 'mapping_quality'], inplace=True)\n",
    "\n",
    "    guide_w_pam = list()\n",
    "\n",
    "    for idx, row in df_a7_2mm_cds.iterrows():\n",
    "        if row['strand'] == '-':\n",
    "            guide_w_pam.append(rc(row['aligned_seq']))\n",
    "        else:\n",
    "            guide_w_pam.append(row['aligned_seq'])\n",
    "\n",
    "    df_a7_2mm_cds['guide_w_pam'] = guide_w_pam\n",
    "    df_a7_2mm_cds['guide'] = df_a7_2mm_cds['guide_w_pam'].str[:-3]\n",
    "\n",
    "    # Create a new column called mismatch_position and extract the mismatch position digits from\n",
    "    # the mismatch column. Then fill NaN values with 0 because NaN means no mismatch.\n",
    "    # Then drop rows where the mismatch occurs after the 10th base.\n",
    "    df_a7_2mm_cds['mismatch'].fillna('Exact Match', inplace=True)\n",
    "    df_a7_2mm_cds = df_a7_2mm_cds[~df_a7_2mm_cds['mismatch'].str.contains('20:')]\n",
    "    df_a7_2mm_cds['num_mutations'] = df_a7_2mm_cds['mismatch'].apply(lambda x: len(x.split(',')))\n",
    "    df_a7_2mm_cds.loc[df_a7_2mm_cds['mismatch'] == 'Exact Match', 'num_mutations'] = 0\n",
    "\n",
    "    df_a7_2mm_cds['num_mutations'] = df_a7_2mm_cds['num_mutations'].astype(int)\n",
    "    \n",
    "    has_no_offtargets = (set(library['sequence']) == set(df_a7_2mm_cds['guide'])) and (len(library['sequence']) == len(df_a7_2mm_cds['guide']))\n",
    "\n",
    "    if not has_no_offtargets:\n",
    "        for _, row in df_a7_2mm_cds[df_a7_2mm_cds['num_mutations'] > 1].iterrows():\n",
    "            idx = library[library['sequence'] == row['guide']].index[0]\n",
    "            two_mm_cds_offtargets[idx] += 1\n",
    "    \n",
    "    for guide in set(df_a7_2mm_cds['guide']):\n",
    "        df_a7_2mm_cds[df_a7_2mm_cds['guide'] == guide].to_csv(f'{species_name}/output/{track}/cds_targets/{guide}_cds_targets.csv', index=False)\n",
    "    \n",
    "    ## ----------\n",
    "    ## Genome 0MM\n",
    "    ## ----------\n",
    "    df_a7_0mm_genomic = pd.read_csv(f'{species_name}/output/{track}/{species_name}_{track}_0mm_genomic_alignment.sam', sep='\\t',\n",
    "                    names=['query_name', 'strand', 'reference_name', 'mapping_position',\n",
    "                        'aligned_seq', 'mapping_quality', 'idk', 'mismatch'])\n",
    "\n",
    "    df_a7_0mm_genomic.drop(columns=['mapping_quality', 'idk'], inplace=True)\n",
    "\n",
    "    guide_w_pam = list()\n",
    "\n",
    "    for idx, row in df_a7_0mm_genomic.iterrows():\n",
    "        if row['strand'] == '-':\n",
    "            guide_w_pam.append(rc(row['aligned_seq']))\n",
    "        else:\n",
    "            guide_w_pam.append(row['aligned_seq'])\n",
    "\n",
    "    df_a7_0mm_genomic['guide_w_pam'] = guide_w_pam\n",
    "    df_a7_0mm_genomic['guide'] = df_a7_0mm_genomic['guide_w_pam'].str[:-3]\n",
    "\n",
    "    # Create a new column called num_mutations and extract the mismatches from\n",
    "    # the mismatch column.\n",
    "    df_a7_0mm_genomic['mismatch'].fillna('Exact Match', inplace=True)\n",
    "    df_a7_0mm_genomic = df_a7_0mm_genomic[~df_a7_0mm_genomic['mismatch'].str.contains('20:')]\n",
    "    df_a7_0mm_genomic['num_mutations'] = df_a7_0mm_genomic['mismatch'].apply(lambda x: len(x.split(',')))\n",
    "    df_a7_0mm_genomic.loc[df_a7_0mm_genomic['mismatch'] == 'Exact Match', 'num_mutations'] = 0\n",
    "\n",
    "    df_a7_0mm_genomic['num_mutations'] = df_a7_0mm_genomic['num_mutations'].astype(int)\n",
    "    \n",
    "    has_no_offtargets = (set(library['sequence']) == set(df_a7_0mm_genomic['guide'])) and (len(library['sequence']) == len(df_a7_0mm_genomic['guide']))\n",
    "\n",
    "    if not has_no_offtargets:\n",
    "        for idx, row in library.iterrows():\n",
    "            seq = row['sequence']\n",
    "            zero_mm_genome_offtargets[idx] += len(df_a7_0mm_genomic[df_a7_0mm_genomic['guide'] == seq]) - 1\n",
    "\n",
    "    ## ----------\n",
    "    ## Genome 1MM\n",
    "    ## ----------\n",
    "    df_a7_1mm_genomic = pd.read_csv(f'{species_name}/output/{track}/{species_name}_{track}_1mm_genomic_alignment.sam', sep='\\t',\n",
    "                names=['query_name', 'strand', 'reference_name', 'mapping_position',\n",
    "                       'aligned_seq', 'mapping_quality', 'idk', 'mismatch'])\n",
    "\n",
    "    df_a7_1mm_genomic.drop(columns=['mapping_quality', 'idk'], inplace=True)\n",
    "\n",
    "    guide_w_pam = list()\n",
    "\n",
    "    for idx, row in df_a7_1mm_genomic.iterrows():\n",
    "        if row['strand'] == '-':\n",
    "            guide_w_pam.append(rc(row['aligned_seq']))\n",
    "        else:\n",
    "            guide_w_pam.append(row['aligned_seq'])\n",
    "\n",
    "    df_a7_1mm_genomic['guide_w_pam'] = guide_w_pam\n",
    "    df_a7_1mm_genomic['guide'] = df_a7_1mm_genomic['guide_w_pam'].str[:-3]\n",
    "\n",
    "    # Create a new column called mismatch_position and extract the mismatch position digits from\n",
    "    # the mismatch column. Then fill NaN values with 0 because NaN means no mismatch.\n",
    "    # Then drop rows where the mismatch occurs after the 10th base.\n",
    "    df_a7_1mm_genomic['mismatch'].fillna('Exact Match', inplace=True)\n",
    "    df_a7_1mm_genomic = df_a7_1mm_genomic[~df_a7_1mm_genomic['mismatch'].str.contains('20:')]\n",
    "    df_a7_1mm_genomic['num_mutations'] = df_a7_1mm_genomic['mismatch'].apply(lambda x: len(x.split(',')))\n",
    "    df_a7_1mm_genomic.loc[df_a7_1mm_genomic['mismatch'] == 'Exact Match', 'num_mutations'] = 0\n",
    "\n",
    "    df_a7_1mm_genomic['num_mutations'] = df_a7_1mm_genomic['num_mutations'].astype(int)\n",
    "    \n",
    "    has_no_offtargets = (set(library['sequence']) == set(df_a7_1mm_genomic['guide'])) and (len(library['sequence']) == len(df_a7_1mm_genomic['guide']))\n",
    "\n",
    "    if not has_no_offtargets:\n",
    "        for _, row in df_a7_1mm_genomic[df_a7_1mm_genomic['num_mutations'] > 0].iterrows():\n",
    "            idx = library[library['sequence'] == row['guide']].index[0]\n",
    "            one_mm_genome_offtargets[idx] += 1\n",
    "\n",
    "    ## ----------\n",
    "    ## Genome 2MM\n",
    "    ## ----------\n",
    "    df_a7_2mm_genomic = pd.read_csv(f'{species_name}/output/{track}/{species_name}_{track}_2mm_genomic_alignment.sam', sep='\\t',\n",
    "                names=['query_name', 'strand', 'reference_name', 'mapping_position',\n",
    "                       'aligned_seq', 'mapping_quality', 'idk', 'mismatch'])\n",
    "\n",
    "    df_a7_2mm_genomic.drop(columns=['mapping_quality', 'idk'], inplace=True)\n",
    "\n",
    "    guide_w_pam = list()\n",
    "\n",
    "    for idx, row in df_a7_2mm_genomic.iterrows():\n",
    "        if row['strand'] == '-':\n",
    "            guide_w_pam.append(rc(row['aligned_seq']))\n",
    "        else:\n",
    "            guide_w_pam.append(row['aligned_seq'])\n",
    "\n",
    "    df_a7_2mm_genomic['guide_w_pam'] = guide_w_pam\n",
    "    df_a7_2mm_genomic['guide'] = df_a7_2mm_genomic['guide_w_pam'].str[:-3]\n",
    "\n",
    "    # Create a new column called num_mutations and extract the mismatches from\n",
    "    # the mismatch column.\n",
    "    df_a7_2mm_genomic['mismatch'].fillna('Exact Match', inplace=True)\n",
    "    df_a7_2mm_genomic = df_a7_2mm_genomic[~df_a7_2mm_genomic['mismatch'].str.contains('20:')]\n",
    "    df_a7_2mm_genomic['num_mutations'] = df_a7_2mm_genomic['mismatch'].apply(lambda x: len(x.split(',')))\n",
    "    df_a7_2mm_genomic.loc[df_a7_2mm_genomic['mismatch'] == 'Exact Match', 'num_mutations'] = 0\n",
    "\n",
    "    df_a7_2mm_genomic['num_mutations'] = df_a7_2mm_genomic['num_mutations'].astype(int)\n",
    "\n",
    "    has_no_offtargets = (set(library['sequence']) == set(df_a7_2mm_genomic['guide'])) and (len(library['sequence']) == len(df_a7_2mm_genomic['guide']))\n",
    "\n",
    "    if not has_no_offtargets:\n",
    "        for _, row in df_a7_2mm_genomic[df_a7_2mm_genomic['num_mutations'] > 1].iterrows():\n",
    "            idx = library[library['sequence'] == row['guide']].index[0]\n",
    "            two_mm_genome_offtargets[idx] += 1\n",
    "\n",
    "    for guide in set(df_a7_2mm_genomic['guide']):\n",
    "        df_a7_2mm_genomic[df_a7_2mm_genomic['guide'] == guide].to_csv(f'{species_name}/output/{track}/genome_targets/{guide}_genome_targets.csv', index=False)\n",
    "    \n",
    "    zero_mm_cds_offtargets = np.array(zero_mm_cds_offtargets)\n",
    "    one_mm_cds_offtargets = np.array(one_mm_cds_offtargets)\n",
    "    two_mm_cds_offtargets = np.array(two_mm_cds_offtargets)\n",
    "\n",
    "    zero_mm_genome_offtargets = np.array(zero_mm_genome_offtargets)\n",
    "    one_mm_genome_offtargets = np.array(one_mm_genome_offtargets)\n",
    "    two_mm_genome_offtargets = np.array(two_mm_genome_offtargets)\n",
    "\n",
    "    library['0mm_cds_offtargets'] = zero_mm_cds_offtargets\n",
    "    library['1mm_cds_offtargets'] = one_mm_cds_offtargets\n",
    "    library['2mm_cds_offtargets'] = two_mm_cds_offtargets\n",
    "\n",
    "    library['0mm_genome_offtargets'] = zero_mm_genome_offtargets\n",
    "    library['1mm_genome_offtargets'] = one_mm_genome_offtargets\n",
    "    library['2mm_genome_offtargets'] = two_mm_genome_offtargets\n",
    "\n",
    "    library.to_csv(f'{species_name}/output/{track}/{track}_lib_offtargets.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Align the augmented library to genome\n",
    "    1. This finds the appropriate NGG PAM guide variants.\n",
    "1. align augmented library to genome, then add a column without PAMS. This column must have the same elements and size as the unaugmented library guides. The size of the alignment will be:\n",
    "\n",
    "1. Equal\n",
    "    1. When either all guides have exactly 1 match (good!), or some guides have NO match (intron event) AND some have more than 1 match (offtarget event, double bad)\n",
    "1. Fewer\n",
    "    1. When some guides have no match (intron event)\n",
    "1. More\n",
    "    1. When some guides have more than 1 match (offtarget)\n",
    "\n",
    "than the library size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "species = ['yarrowia_lipolytica', 'kluyveromyces_marxianus',\n",
    "           'komagataella_phaffii', 'saccharomyces_cerevisiae']\n",
    "\n",
    "for track in ['a7', 'e1']:\n",
    "    for s in species:\n",
    "        GO(s, track)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>reference_name</th>\n",
       "      <th>mapping_position</th>\n",
       "      <th>aligned_seq</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>guide_w_pam</th>\n",
       "      <th>guide</th>\n",
       "      <th>num_mutations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>READ_1</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_006071.1</td>\n",
       "      <td>1265190</td>\n",
       "      <td>CCTTCTGTCATGAACGTCGTCAT</td>\n",
       "      <td>5:G&gt;C,17:G&gt;T</td>\n",
       "      <td>ATGACGACGTTCATGACAGAAGG</td>\n",
       "      <td>ATGACGACGTTCATGACAGA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>READ_3</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_006068.1</td>\n",
       "      <td>2177351</td>\n",
       "      <td>CCATCTGTCATGAACGTCGTCAT</td>\n",
       "      <td>Exact Match</td>\n",
       "      <td>ATGACGACGTTCATGACAGATGG</td>\n",
       "      <td>ATGACGACGTTCATGACAGA</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>READ_4</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_006072.1</td>\n",
       "      <td>2649613</td>\n",
       "      <td>CCCTCTGTCATGAACGTCGTCAT</td>\n",
       "      <td>5:G&gt;C,17:C&gt;T</td>\n",
       "      <td>ATGACGACGTTCATGACAGAGGG</td>\n",
       "      <td>ATGACGACGTTCATGACAGA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>READ_6</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_006070.1</td>\n",
       "      <td>3355366</td>\n",
       "      <td>CCGATCCACTTGGTGGCAGAGTG</td>\n",
       "      <td>Exact Match</td>\n",
       "      <td>CACTCTGCCACCAAGTGGATCGG</td>\n",
       "      <td>CACTCTGCCACCAAGTGGAT</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>READ_10</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_006068.1</td>\n",
       "      <td>2176511</td>\n",
       "      <td>CATCTCCAGATGATTGCCATCGG</td>\n",
       "      <td>Exact Match</td>\n",
       "      <td>CATCTCCAGATGATTGCCATCGG</td>\n",
       "      <td>CATCTCCAGATGATTGCCAT</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>READ_10</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_006068.1</td>\n",
       "      <td>2611782</td>\n",
       "      <td>CCGATGGCAATCATCTGGAGATG</td>\n",
       "      <td>5:A&gt;G,17:A&gt;G</td>\n",
       "      <td>CATCTCCAGATGATTGCCATCGG</td>\n",
       "      <td>CATCTCCAGATGATTGCCAT</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>READ_11</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_006069.1</td>\n",
       "      <td>2426679</td>\n",
       "      <td>CATCTCCAGATGATTGCCATTGG</td>\n",
       "      <td>17:T&gt;C</td>\n",
       "      <td>CATCTCCAGATGATTGCCATTGG</td>\n",
       "      <td>CATCTCCAGATGATTGCCAT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>READ_11</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_006069.1</td>\n",
       "      <td>1355833</td>\n",
       "      <td>CCAATGGCAATCATCTGGAGATG</td>\n",
       "      <td>5:C&gt;G</td>\n",
       "      <td>CATCTCCAGATGATTGCCATTGG</td>\n",
       "      <td>CATCTCCAGATGATTGCCAT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>READ_11</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_006068.1</td>\n",
       "      <td>2577048</td>\n",
       "      <td>CATCTCCAGATGATTGCCATTGG</td>\n",
       "      <td>5:G&gt;C</td>\n",
       "      <td>CATCTCCAGATGATTGCCATTGG</td>\n",
       "      <td>CATCTCCAGATGATTGCCAT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>READ_16</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_006071.1</td>\n",
       "      <td>735514</td>\n",
       "      <td>CCCTGGATGGCTGAGACCTTTGG</td>\n",
       "      <td>Exact Match</td>\n",
       "      <td>CCAAAGGTCTCAGCCATCCAGGG</td>\n",
       "      <td>CCAAAGGTCTCAGCCATCCA</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>READ_17</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_006068.1</td>\n",
       "      <td>2176514</td>\n",
       "      <td>CTCCAGATGATTGCCATCGGAGG</td>\n",
       "      <td>Exact Match</td>\n",
       "      <td>CTCCAGATGATTGCCATCGGAGG</td>\n",
       "      <td>CTCCAGATGATTGCCATCGG</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>READ_17</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_006068.1</td>\n",
       "      <td>2577051</td>\n",
       "      <td>CTCCAGATGATTGCCATCGGAGG</td>\n",
       "      <td>2:G&gt;C,17:T&gt;C</td>\n",
       "      <td>CTCCAGATGATTGCCATCGGAGG</td>\n",
       "      <td>CTCCAGATGATTGCCATCGG</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>READ_18</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_006069.1</td>\n",
       "      <td>2426682</td>\n",
       "      <td>CTCCAGATGATTGCCATCGGCGG</td>\n",
       "      <td>14:T&gt;C,17:T&gt;C</td>\n",
       "      <td>CTCCAGATGATTGCCATCGGCGG</td>\n",
       "      <td>CTCCAGATGATTGCCATCGG</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>READ_19</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_006069.1</td>\n",
       "      <td>1355830</td>\n",
       "      <td>CCACCGATGGCAATCATCTGGAG</td>\n",
       "      <td>2:C&gt;G,17:A&gt;G</td>\n",
       "      <td>CTCCAGATGATTGCCATCGGTGG</td>\n",
       "      <td>CTCCAGATGATTGCCATCGG</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>READ_19</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_006068.1</td>\n",
       "      <td>2611779</td>\n",
       "      <td>CCACCGATGGCAATCATCTGGAG</td>\n",
       "      <td>2:A&gt;G,14:A&gt;G</td>\n",
       "      <td>CTCCAGATGATTGCCATCGGTGG</td>\n",
       "      <td>CTCCAGATGATTGCCATCGG</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>READ_21</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_006071.1</td>\n",
       "      <td>736288</td>\n",
       "      <td>GATGACCAGGTGAAGATTCGAGG</td>\n",
       "      <td>Exact Match</td>\n",
       "      <td>GATGACCAGGTGAAGATTCGAGG</td>\n",
       "      <td>GATGACCAGGTGAAGATTCG</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>READ_25</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_006070.1</td>\n",
       "      <td>3355363</td>\n",
       "      <td>CCTCCGATCCACTTGGTGGCAGA</td>\n",
       "      <td>Exact Match</td>\n",
       "      <td>TCTGCCACCAAGTGGATCGGAGG</td>\n",
       "      <td>TCTGCCACCAAGTGGATCGG</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   query_name strand reference_name  mapping_position  \\\n",
       "1      READ_1      -    NC_006071.1           1265190   \n",
       "3      READ_3      -    NC_006068.1           2177351   \n",
       "5      READ_4      -    NC_006072.1           2649613   \n",
       "7      READ_6      -    NC_006070.1           3355366   \n",
       "14    READ_10      +    NC_006068.1           2176511   \n",
       "17    READ_10      -    NC_006068.1           2611782   \n",
       "20    READ_11      +    NC_006069.1           2426679   \n",
       "21    READ_11      -    NC_006069.1           1355833   \n",
       "22    READ_11      +    NC_006068.1           2577048   \n",
       "30    READ_16      -    NC_006071.1            735514   \n",
       "31    READ_17      +    NC_006068.1           2176514   \n",
       "32    READ_17      +    NC_006068.1           2577051   \n",
       "34    READ_18      +    NC_006069.1           2426682   \n",
       "36    READ_19      -    NC_006069.1           1355830   \n",
       "37    READ_19      -    NC_006068.1           2611779   \n",
       "39    READ_21      +    NC_006071.1            736288   \n",
       "43    READ_25      -    NC_006070.1           3355363   \n",
       "\n",
       "                aligned_seq       mismatch              guide_w_pam  \\\n",
       "1   CCTTCTGTCATGAACGTCGTCAT   5:G>C,17:G>T  ATGACGACGTTCATGACAGAAGG   \n",
       "3   CCATCTGTCATGAACGTCGTCAT    Exact Match  ATGACGACGTTCATGACAGATGG   \n",
       "5   CCCTCTGTCATGAACGTCGTCAT   5:G>C,17:C>T  ATGACGACGTTCATGACAGAGGG   \n",
       "7   CCGATCCACTTGGTGGCAGAGTG    Exact Match  CACTCTGCCACCAAGTGGATCGG   \n",
       "14  CATCTCCAGATGATTGCCATCGG    Exact Match  CATCTCCAGATGATTGCCATCGG   \n",
       "17  CCGATGGCAATCATCTGGAGATG   5:A>G,17:A>G  CATCTCCAGATGATTGCCATCGG   \n",
       "20  CATCTCCAGATGATTGCCATTGG         17:T>C  CATCTCCAGATGATTGCCATTGG   \n",
       "21  CCAATGGCAATCATCTGGAGATG          5:C>G  CATCTCCAGATGATTGCCATTGG   \n",
       "22  CATCTCCAGATGATTGCCATTGG          5:G>C  CATCTCCAGATGATTGCCATTGG   \n",
       "30  CCCTGGATGGCTGAGACCTTTGG    Exact Match  CCAAAGGTCTCAGCCATCCAGGG   \n",
       "31  CTCCAGATGATTGCCATCGGAGG    Exact Match  CTCCAGATGATTGCCATCGGAGG   \n",
       "32  CTCCAGATGATTGCCATCGGAGG   2:G>C,17:T>C  CTCCAGATGATTGCCATCGGAGG   \n",
       "34  CTCCAGATGATTGCCATCGGCGG  14:T>C,17:T>C  CTCCAGATGATTGCCATCGGCGG   \n",
       "36  CCACCGATGGCAATCATCTGGAG   2:C>G,17:A>G  CTCCAGATGATTGCCATCGGTGG   \n",
       "37  CCACCGATGGCAATCATCTGGAG   2:A>G,14:A>G  CTCCAGATGATTGCCATCGGTGG   \n",
       "39  GATGACCAGGTGAAGATTCGAGG    Exact Match  GATGACCAGGTGAAGATTCGAGG   \n",
       "43  CCTCCGATCCACTTGGTGGCAGA    Exact Match  TCTGCCACCAAGTGGATCGGAGG   \n",
       "\n",
       "                   guide  num_mutations  \n",
       "1   ATGACGACGTTCATGACAGA              2  \n",
       "3   ATGACGACGTTCATGACAGA              0  \n",
       "5   ATGACGACGTTCATGACAGA              2  \n",
       "7   CACTCTGCCACCAAGTGGAT              0  \n",
       "14  CATCTCCAGATGATTGCCAT              0  \n",
       "17  CATCTCCAGATGATTGCCAT              2  \n",
       "20  CATCTCCAGATGATTGCCAT              1  \n",
       "21  CATCTCCAGATGATTGCCAT              1  \n",
       "22  CATCTCCAGATGATTGCCAT              1  \n",
       "30  CCAAAGGTCTCAGCCATCCA              0  \n",
       "31  CTCCAGATGATTGCCATCGG              0  \n",
       "32  CTCCAGATGATTGCCATCGG              2  \n",
       "34  CTCCAGATGATTGCCATCGG              2  \n",
       "36  CTCCAGATGATTGCCATCGG              2  \n",
       "37  CTCCAGATGATTGCCATCGG              2  \n",
       "39  GATGACCAGGTGAAGATTCG              0  \n",
       "43  TCTGCCACCAAGTGGATCGG              0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "species_name = 'yarrowia_lipolytica'\n",
    "track = 'a7'\n",
    "\n",
    "library = pd.read_csv(f'{species_name}/{species_name}_{track}_library.csv')\n",
    "\n",
    "zero_mm_cds_offtargets = [0] * len(library)\n",
    "one_mm_cds_offtargets = [0] * len(library)\n",
    "two_mm_cds_offtargets = [0] * len(library)\n",
    "\n",
    "zero_mm_genome_offtargets = [0] * len(library)\n",
    "one_mm_genome_offtargets = [0] * len(library)\n",
    "two_mm_genome_offtargets = [0] * len(library)\n",
    "\n",
    "\n",
    "df_a7_2mm_genomic = pd.read_csv(f'{species_name}/output/{track}/{species_name}_{track}_2mm_genomic_alignment.sam', sep='\\t',\n",
    "                names=['query_name', 'strand', 'reference_name', 'mapping_position',\n",
    "                       'aligned_seq', 'mapping_quality', 'idk', 'mismatch'])\n",
    "\n",
    "df_a7_2mm_genomic.drop(columns=['mapping_quality', 'idk'], inplace=True)\n",
    "\n",
    "guide_w_pam = list()\n",
    "\n",
    "for idx, row in df_a7_2mm_genomic.iterrows():\n",
    "    if row['strand'] == '-':\n",
    "        guide_w_pam.append(rc(row['aligned_seq']))\n",
    "    else:\n",
    "        guide_w_pam.append(row['aligned_seq'])\n",
    "\n",
    "df_a7_2mm_genomic['guide_w_pam'] = guide_w_pam\n",
    "df_a7_2mm_genomic['guide'] = df_a7_2mm_genomic['guide_w_pam'].str[:-3]\n",
    "\n",
    "# Create a new column called num_mutations and extract the mismatches from\n",
    "# the mismatch column.\n",
    "df_a7_2mm_genomic['mismatch'].fillna('Exact Match', inplace=True)\n",
    "df_a7_2mm_genomic = df_a7_2mm_genomic[~df_a7_2mm_genomic['mismatch'].str.contains('20:')]\n",
    "df_a7_2mm_genomic['num_mutations'] = df_a7_2mm_genomic['mismatch'].apply(lambda x: len(x.split(',')))\n",
    "df_a7_2mm_genomic.loc[df_a7_2mm_genomic['mismatch'] == 'Exact Match', 'num_mutations'] = 0\n",
    "\n",
    "df_a7_2mm_genomic['num_mutations'] = df_a7_2mm_genomic['num_mutations'].astype(int)\n",
    "\n",
    "df_a7_2mm_genomic\n",
    "\n",
    "# has_no_offtargets = (set(library['sequence']) == set(df_a7_2mm_genomic['guide'])) and (len(library['sequence']) == len(df_a7_2mm_genomic['guide']))\n",
    "\n",
    "# if not has_no_offtargets:\n",
    "#     for _, row in df_a7_2mm_genomic[df_a7_2mm_genomic['num_mutations'] > 1].iterrows():\n",
    "#         idx = library[library['sequence'] == row['guide']].index[0]\n",
    "#         two_mm_genome_offtargets[idx] += 1"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "allegro",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
