{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Protocol for checking CDS offtargets:\n",
    "\n",
    "1. Get the augmented library guides (guides + 4 PAMs) and bowtie align them to CDS\n",
    "    1. This finds the appropriate NGG PAM guide variants\n",
    "    1. Two things can be observed from the alignment output:\n",
    "        1. For 0mm, which guide + PAM sequences have exact matches and their locations\n",
    "            1. If a guide has more than 1 exact match with the same or different PAM, those are off targets.\n",
    "        1. For 1mm, there are alignments that indicate 1 mismatch but in the N of the PAM region. I throw these out. Because these target the same region as the \"original\" guide + PAM with which they mismatch once. If they targeted another region with 1 mismatch, there would have been a correct guide + PAM variant in the previous step (0mm) which would have been recorded already.\n",
    "        1. For 2mm, there are alignments that indicate 1 mismatch in the N of the PAM, and 1 mismatch somewhere else in the sequence. I throw these out. Because the correct variant and 1 mismatch would have been recorded in the previous step (1mm)\n",
    "1. The number of exact matches of 0MM CDS must be the same as the library size.\n",
    "    1. Actually, size is a bad measure. What if the same guide with different PAMs has matches?\n",
    "1. Need to align augmented library to CDS, then add a column without PAMS. This column must have the same elements and size as the unaugmented library guides. The size of the alignment will be equal or more (w/ offtargets) than the library."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "if there is an intron separation event, we can't see it in the protocol above. Need to check genome.\n",
    "\n",
    "1. Align the augmented library to genome\n",
    "    1. This finds the appropriate NGG PAM guide variants.\n",
    "1. align augmented library to genome, then add a column without PAMS. This column must have the same elements and size as the unaugmented library guides. The size of the alignment will be:\n",
    "\n",
    "1. Equal\n",
    "    1. When either all guides have exactly 1 match (good!), or some guides have NO match (intron event) AND some have more than 1 match (offtarget event, double bad)\n",
    "1. Fewer\n",
    "    1. When some guides have no match (intron event)\n",
    "1. More\n",
    "    1. When some guides have more than 1 match (offtarget)\n",
    "\n",
    "than the library size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "def rc(string):\n",
    "    s = ''\n",
    "    for c in string:\n",
    "        if c == 'A': s += 'T'\n",
    "        elif c == 'C': s += 'G'\n",
    "        elif c == 'G': s += 'C'\n",
    "        elif c == 'T': s += 'A'\n",
    "    return s[::-1]\n",
    "\n",
    "def GO(species_name, track):\n",
    "    if not os.path.exists(f'{species_name}/output/{track}/cds_targets'):\n",
    "        os.makedirs(f'{species_name}/output/{track}/cds_targets')\n",
    "    \n",
    "    if not os.path.exists(f'{species_name}/output/{track}/genome_targets'):\n",
    "        os.makedirs(f'{species_name}/output/{track}/genome_targets')\n",
    "\n",
    "    library = pd.read_csv(f'{species_name}/{species_name}_{track}_library.csv')\n",
    "\n",
    "    zero_mm_cds_offtargets = [0] * len(library)\n",
    "    one_mm_cds_offtargets = [0] * len(library)\n",
    "    two_mm_cds_offtargets = [0] * len(library)\n",
    "\n",
    "    zero_mm_genome_offtargets = [0] * len(library)\n",
    "    one_mm_genome_offtargets = [0] * len(library)\n",
    "    two_mm_genome_offtargets = [0] * len(library)\n",
    "\n",
    "\n",
    "    df_a7_0mm_cds = pd.read_csv(f'{species_name}/output/{track}/{species_name}_{track}_0mm_cds_alignment.sam', sep='\\t',\n",
    "                names=['query_name', 'strand', 'reference_name', 'mapping_position',\n",
    "                       'aligned_seq', 'mapping_quality', 'idk', 'mismatch'])\n",
    "\n",
    "    df_a7_0mm_cds.drop(columns=['idk', 'mapping_quality'], inplace=True)\n",
    "\n",
    "    guide_w_pam = list()\n",
    "\n",
    "    # Add a reverse complement column if strand is negative - otherwise write the same sequence\n",
    "    for idx, row in df_a7_0mm_cds.iterrows():\n",
    "        if row['strand'] == '-':\n",
    "            guide_w_pam.append(rc(row['aligned_seq']))\n",
    "        else:\n",
    "            guide_w_pam.append(row['aligned_seq'])\n",
    "\n",
    "    df_a7_0mm_cds['guide_w_pam'] = guide_w_pam\n",
    "    df_a7_0mm_cds['guide'] = df_a7_0mm_cds['guide_w_pam'].str[:-3]\n",
    "\n",
    "    df_a7_0mm_cds['mismatch'].fillna('Exact Match', inplace=True)\n",
    "    df_a7_0mm_cds = df_a7_0mm_cds[~df_a7_0mm_cds['mismatch'].str.contains('20:')]\n",
    "    df_a7_0mm_cds['num_mutations'] = df_a7_0mm_cds['mismatch'].apply(lambda x: len(x.split(',')))\n",
    "    df_a7_0mm_cds.loc[df_a7_0mm_cds['mismatch'] == 'Exact Match', 'num_mutations'] = 0\n",
    "\n",
    "    df_a7_0mm_cds['num_mutations'] = df_a7_0mm_cds['num_mutations'].astype(int)\n",
    "\n",
    "    has_no_offtargets = (set(library['sequence']) == set(df_a7_0mm_cds['guide'])) and (len(library['sequence']) == len(df_a7_0mm_cds['guide']))\n",
    "\n",
    "    if not has_no_offtargets:\n",
    "        for idx, row in library.iterrows():\n",
    "            seq = row['sequence']\n",
    "            zero_mm_cds_offtargets[idx] += len(df_a7_0mm_cds[df_a7_0mm_cds['guide'] == seq]) - 1\n",
    "\n",
    "    ## -------\n",
    "    ## CDS 1MM\n",
    "    ## -------\n",
    "    df_a7_1mm_cds = pd.read_csv(f'{species_name}/output/{track}/{species_name}_{track}_1mm_cds_alignment.sam', sep='\\t',\n",
    "                names=['query_name', 'strand', 'reference_name', 'mapping_position',\n",
    "                       'aligned_seq', 'mapping_quality', 'idk', 'mismatch'])\n",
    "\n",
    "    df_a7_1mm_cds.drop(columns=['idk', 'mapping_quality'], inplace=True)\n",
    "\n",
    "    guide_w_pam = list()\n",
    "\n",
    "    for idx, row in df_a7_1mm_cds.iterrows():\n",
    "        if row['strand'] == '-':\n",
    "            guide_w_pam.append(rc(row['aligned_seq']))\n",
    "        else:\n",
    "            guide_w_pam.append(row['aligned_seq'])\n",
    "\n",
    "    df_a7_1mm_cds['guide_w_pam'] = guide_w_pam\n",
    "    df_a7_1mm_cds['guide'] = df_a7_1mm_cds['guide_w_pam'].str[:-3]\n",
    "\n",
    "    # Create a new column called num_mutations and extract the mismatches from\n",
    "    # the mismatch column.\n",
    "    df_a7_1mm_cds['mismatch'].fillna('Exact Match', inplace=True)\n",
    "    df_a7_1mm_cds = df_a7_1mm_cds[~df_a7_1mm_cds['mismatch'].str.contains('20:')]\n",
    "    df_a7_1mm_cds['num_mutations'] = df_a7_1mm_cds['mismatch'].apply(lambda x: len(x.split(',')))\n",
    "    df_a7_1mm_cds.loc[df_a7_1mm_cds['mismatch'] == 'Exact Match', 'num_mutations'] = 0\n",
    "\n",
    "    df_a7_1mm_cds['num_mutations'] = df_a7_1mm_cds['num_mutations'].astype(int)\n",
    "    \n",
    "    has_no_offtargets = (set(library['sequence']) == set(df_a7_1mm_cds['guide'])) and (len(library['sequence']) == len(df_a7_1mm_cds['guide']))\n",
    "\n",
    "    if not has_no_offtargets:\n",
    "        for _, row in df_a7_1mm_cds[df_a7_1mm_cds['num_mutations'] > 0].iterrows():\n",
    "            idx = library[library['sequence'] == row['guide']].index[0]\n",
    "            one_mm_cds_offtargets[idx] += 1\n",
    "\n",
    "    ## ------\n",
    "    # CDS 2MM\n",
    "    ## ------\n",
    "    df_a7_2mm_cds = pd.read_csv(f'{species_name}/output/{track}/{species_name}_{track}_2mm_cds_alignment.sam', sep='\\t',\n",
    "                names=['query_name', 'strand', 'reference_name', 'mapping_position',\n",
    "                       'aligned_seq', 'mapping_quality', 'idk', 'mismatch'])\n",
    "\n",
    "    df_a7_2mm_cds.drop(columns=['idk', 'mapping_quality'], inplace=True)\n",
    "\n",
    "    guide_w_pam = list()\n",
    "\n",
    "    for idx, row in df_a7_2mm_cds.iterrows():\n",
    "        if row['strand'] == '-':\n",
    "            guide_w_pam.append(rc(row['aligned_seq']))\n",
    "        else:\n",
    "            guide_w_pam.append(row['aligned_seq'])\n",
    "\n",
    "    df_a7_2mm_cds['guide_w_pam'] = guide_w_pam\n",
    "    df_a7_2mm_cds['guide'] = df_a7_2mm_cds['guide_w_pam'].str[:-3]\n",
    "\n",
    "    # Create a new column called mismatch_position and extract the mismatch position digits from\n",
    "    # the mismatch column. Then fill NaN values with 0 because NaN means no mismatch.\n",
    "    # Then drop rows where the mismatch occurs after the 10th base.\n",
    "    df_a7_2mm_cds['mismatch'].fillna('Exact Match', inplace=True)\n",
    "    df_a7_2mm_cds = df_a7_2mm_cds[~df_a7_2mm_cds['mismatch'].str.contains('20:')]\n",
    "    df_a7_2mm_cds['num_mutations'] = df_a7_2mm_cds['mismatch'].apply(lambda x: len(x.split(',')))\n",
    "    df_a7_2mm_cds.loc[df_a7_2mm_cds['mismatch'] == 'Exact Match', 'num_mutations'] = 0\n",
    "\n",
    "    df_a7_2mm_cds['num_mutations'] = df_a7_2mm_cds['num_mutations'].astype(int)\n",
    "    \n",
    "    has_no_offtargets = (set(library['sequence']) == set(df_a7_2mm_cds['guide'])) and (len(library['sequence']) == len(df_a7_2mm_cds['guide']))\n",
    "\n",
    "    if not has_no_offtargets:\n",
    "        for _, row in df_a7_2mm_cds[df_a7_2mm_cds['num_mutations'] > 1].iterrows():\n",
    "            idx = library[library['sequence'] == row['guide']].index[0]\n",
    "            two_mm_cds_offtargets[idx] += 1\n",
    "    \n",
    "    for guide in set(df_a7_2mm_cds['guide']):\n",
    "        df_a7_2mm_cds[df_a7_2mm_cds['guide'] == guide].to_csv(f'{species_name}/output/{track}/cds_targets/{guide}_cds_targets.csv', index=False)\n",
    "    \n",
    "    ## ----------\n",
    "    ## Genome 0MM\n",
    "    ## ----------\n",
    "    df_a7_0mm_genomic = pd.read_csv(f'{species_name}/output/{track}/{species_name}_{track}_0mm_genomic_alignment.sam', sep='\\t',\n",
    "                    names=['query_name', 'strand', 'reference_name', 'mapping_position',\n",
    "                        'aligned_seq', 'mapping_quality', 'idk', 'mismatch'])\n",
    "\n",
    "    df_a7_0mm_genomic.drop(columns=['mapping_quality', 'idk'], inplace=True)\n",
    "\n",
    "    guide_w_pam = list()\n",
    "\n",
    "    for idx, row in df_a7_0mm_genomic.iterrows():\n",
    "        if row['strand'] == '-':\n",
    "            guide_w_pam.append(rc(row['aligned_seq']))\n",
    "        else:\n",
    "            guide_w_pam.append(row['aligned_seq'])\n",
    "\n",
    "    df_a7_0mm_genomic['guide_w_pam'] = guide_w_pam\n",
    "    df_a7_0mm_genomic['guide'] = df_a7_0mm_genomic['guide_w_pam'].str[:-3]\n",
    "\n",
    "    # Create a new column called num_mutations and extract the mismatches from\n",
    "    # the mismatch column.\n",
    "    df_a7_0mm_genomic['mismatch'].fillna('Exact Match', inplace=True)\n",
    "    df_a7_0mm_genomic = df_a7_0mm_genomic[~df_a7_0mm_genomic['mismatch'].str.contains('20:')]\n",
    "    df_a7_0mm_genomic['num_mutations'] = df_a7_0mm_genomic['mismatch'].apply(lambda x: len(x.split(',')))\n",
    "    df_a7_0mm_genomic.loc[df_a7_0mm_genomic['mismatch'] == 'Exact Match', 'num_mutations'] = 0\n",
    "\n",
    "    df_a7_0mm_genomic['num_mutations'] = df_a7_0mm_genomic['num_mutations'].astype(int)\n",
    "    \n",
    "    has_no_offtargets = (set(library['sequence']) == set(df_a7_0mm_genomic['guide'])) and (len(library['sequence']) == len(df_a7_0mm_genomic['guide']))\n",
    "\n",
    "    if not has_no_offtargets:\n",
    "        for idx, row in library.iterrows():\n",
    "            seq = row['sequence']\n",
    "            zero_mm_genome_offtargets[idx] += len(df_a7_0mm_genomic[df_a7_0mm_genomic['guide'] == seq]) - 1\n",
    "\n",
    "    ## ----------\n",
    "    ## Genome 1MM\n",
    "    ## ----------\n",
    "    df_a7_1mm_genomic = pd.read_csv(f'{species_name}/output/{track}/{species_name}_{track}_1mm_genomic_alignment.sam', sep='\\t',\n",
    "                names=['query_name', 'strand', 'reference_name', 'mapping_position',\n",
    "                       'aligned_seq', 'mapping_quality', 'idk', 'mismatch'])\n",
    "\n",
    "    df_a7_1mm_genomic.drop(columns=['mapping_quality', 'idk'], inplace=True)\n",
    "\n",
    "    guide_w_pam = list()\n",
    "\n",
    "    for idx, row in df_a7_1mm_genomic.iterrows():\n",
    "        if row['strand'] == '-':\n",
    "            guide_w_pam.append(rc(row['aligned_seq']))\n",
    "        else:\n",
    "            guide_w_pam.append(row['aligned_seq'])\n",
    "\n",
    "    df_a7_1mm_genomic['guide_w_pam'] = guide_w_pam\n",
    "    df_a7_1mm_genomic['guide'] = df_a7_1mm_genomic['guide_w_pam'].str[:-3]\n",
    "\n",
    "    # Create a new column called mismatch_position and extract the mismatch position digits from\n",
    "    # the mismatch column. Then fill NaN values with 0 because NaN means no mismatch.\n",
    "    # Then drop rows where the mismatch occurs after the 10th base.\n",
    "    df_a7_1mm_genomic['mismatch'].fillna('Exact Match', inplace=True)\n",
    "    df_a7_1mm_genomic = df_a7_1mm_genomic[~df_a7_1mm_genomic['mismatch'].str.contains('20:')]\n",
    "    df_a7_1mm_genomic['num_mutations'] = df_a7_1mm_genomic['mismatch'].apply(lambda x: len(x.split(',')))\n",
    "    df_a7_1mm_genomic.loc[df_a7_1mm_genomic['mismatch'] == 'Exact Match', 'num_mutations'] = 0\n",
    "\n",
    "    df_a7_1mm_genomic['num_mutations'] = df_a7_1mm_genomic['num_mutations'].astype(int)\n",
    "    \n",
    "    has_no_offtargets = (set(library['sequence']) == set(df_a7_1mm_genomic['guide'])) and (len(library['sequence']) == len(df_a7_1mm_genomic['guide']))\n",
    "\n",
    "    if not has_no_offtargets:\n",
    "        for _, row in df_a7_1mm_genomic[df_a7_1mm_genomic['num_mutations'] > 0].iterrows():\n",
    "            idx = library[library['sequence'] == row['guide']].index[0]\n",
    "            one_mm_genome_offtargets[idx] += 1\n",
    "\n",
    "    ## ----------\n",
    "    ## Genome 2MM\n",
    "    ## ----------\n",
    "    df_a7_2mm_genomic = pd.read_csv(f'{species_name}/output/{track}/{species_name}_{track}_2mm_genomic_alignment.sam', sep='\\t',\n",
    "                names=['query_name', 'strand', 'reference_name', 'mapping_position',\n",
    "                       'aligned_seq', 'mapping_quality', 'idk', 'mismatch'])\n",
    "\n",
    "    df_a7_2mm_genomic.drop(columns=['mapping_quality', 'idk'], inplace=True)\n",
    "\n",
    "    guide_w_pam = list()\n",
    "\n",
    "    for idx, row in df_a7_2mm_genomic.iterrows():\n",
    "        if row['strand'] == '-':\n",
    "            guide_w_pam.append(rc(row['aligned_seq']))\n",
    "        else:\n",
    "            guide_w_pam.append(row['aligned_seq'])\n",
    "\n",
    "    df_a7_2mm_genomic['guide_w_pam'] = guide_w_pam\n",
    "    df_a7_2mm_genomic['guide'] = df_a7_2mm_genomic['guide_w_pam'].str[:-3]\n",
    "\n",
    "    # Create a new column called num_mutations and extract the mismatches from\n",
    "    # the mismatch column.\n",
    "    df_a7_2mm_genomic['mismatch'].fillna('Exact Match', inplace=True)\n",
    "    df_a7_2mm_genomic = df_a7_2mm_genomic[~df_a7_2mm_genomic['mismatch'].str.contains('20:')]\n",
    "    df_a7_2mm_genomic['num_mutations'] = df_a7_2mm_genomic['mismatch'].apply(lambda x: len(x.split(',')))\n",
    "    df_a7_2mm_genomic.loc[df_a7_2mm_genomic['mismatch'] == 'Exact Match', 'num_mutations'] = 0\n",
    "\n",
    "    df_a7_2mm_genomic['num_mutations'] = df_a7_2mm_genomic['num_mutations'].astype(int)\n",
    "\n",
    "    has_no_offtargets = (set(library['sequence']) == set(df_a7_2mm_genomic['guide'])) and (len(library['sequence']) == len(df_a7_2mm_genomic['guide']))\n",
    "\n",
    "    if not has_no_offtargets:\n",
    "        for _, row in df_a7_2mm_genomic[df_a7_2mm_genomic['num_mutations'] > 1].iterrows():\n",
    "            idx = library[library['sequence'] == row['guide']].index[0]\n",
    "            two_mm_genome_offtargets[idx] += 1\n",
    "\n",
    "    for guide in set(df_a7_2mm_genomic['guide']):\n",
    "        df_a7_2mm_genomic[df_a7_2mm_genomic['guide'] == guide].to_csv(f'{species_name}/output/{track}/genome_targets/{guide}_genome_targets.csv', index=False)\n",
    "    \n",
    "    zero_mm_cds_offtargets = np.array(zero_mm_cds_offtargets)\n",
    "    one_mm_cds_offtargets = np.array(one_mm_cds_offtargets)\n",
    "    two_mm_cds_offtargets = np.array(two_mm_cds_offtargets)\n",
    "\n",
    "    zero_mm_genome_offtargets = np.array(zero_mm_genome_offtargets)\n",
    "    one_mm_genome_offtargets = np.array(one_mm_genome_offtargets)\n",
    "    two_mm_genome_offtargets = np.array(two_mm_genome_offtargets)\n",
    "\n",
    "    library['0mm_cds_offtargets'] = zero_mm_cds_offtargets\n",
    "    library['1mm_cds_offtargets'] = one_mm_cds_offtargets\n",
    "    library['2mm_cds_offtargets'] = two_mm_cds_offtargets\n",
    "\n",
    "    library['0mm_genome_offtargets'] = zero_mm_genome_offtargets\n",
    "    library['1mm_genome_offtargets'] = one_mm_genome_offtargets\n",
    "    library['2mm_genome_offtargets'] = two_mm_genome_offtargets\n",
    "\n",
    "    library.to_csv(f'{species_name}/output/{track}/{track}_lib_offtargets.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Align the augmented library to genome\n",
    "    1. This finds the appropriate NGG PAM guide variants.\n",
    "1. align augmented library to genome, then add a column without PAMS. This column must have the same elements and size as the unaugmented library guides. The size of the alignment will be:\n",
    "\n",
    "1. Equal\n",
    "    1. When either all guides have exactly 1 match (good!), or some guides have NO match (intron event) AND some have more than 1 match (offtarget event, double bad)\n",
    "1. Fewer\n",
    "    1. When some guides have no match (intron event)\n",
    "1. More\n",
    "    1. When some guides have more than 1 match (offtarget)\n",
    "\n",
    "than the library size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "species = ['yarrowia_lipolytica', 'kluyveromyces_marxianus',\n",
    "           'komagataella_phaffii', 'saccharomyces_cerevisiae']\n",
    "\n",
    "for track in ['a7', 'e1']:\n",
    "    for s in species:\n",
    "        GO(s, track)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "allegro",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
